### PR TITLE
WIP: Fix dotnetcore

### PR DESCRIPTION
**Pull request by matthid**
This should remove various hacks added over the time when I was kind of offline.

TODO:
 - [ ] proper runtime support in InstallModel
 - [x] disabled "simplify" code-path for msbuild targets -> this part was reviewed, I think the current behavior is more deterministic and if we simplify conditions for targets we should do the same for references... 
 
Breaks:
 - ~~`/Lib/...` is lo longer supported (folder in nuget package needs to be `/lib/...`~~ works again - feels wrong
 - Currently runtime support is removed (see above)

Adds:
 - `InstallModel` can properly resolve reference and runtime assemblies (required in FAKE vNext for compilation and running of scripts)
 - This enables the complete nuget replacement in dotnetcore world at a later time.
 - Incidentally Improves C++ support

### CODE REVIEW

File: src/Paket.Core/Utils.fs [hunk 1]
```diff
@@ -158,60 +158,6 @@ let inline createRelativePath root path =
...
-        else
-            if infix = "runtimes" then
-                Some("runtimes" + getNative path)
-            elif infix = "ref" then
-                let libPart = path.Substring(startPos + infix.Length + 1, endPos - startPos - infix.Length - 1)
-                Some libPart         
-            else
-                let nativePart = getNative path
-                let libPart = path.Substring(startPos + infix.Length + 1, endPos - startPos - infix.Length - 1)
-                Some (libPart + nativePart)
```
Comment (matthid) [inline src/Paket.Core/Utils.fs]: Removing this code is really gives one a good feeling :)

File: src/Paket.Core/Files/ProjectFile.fs [hunk 2]
```diff
@@ -788,16 +796,17 @@ module ProjectFile =
...
         let globalTargets, frameworkSpecificTargets =
             if not importTargets then List.empty, List.empty else
-            let sortedTargets = model.TargetsFileFolders |> List.sortBy (fun lib -> lib.Name)
-            sortedTargets
-            |> List.partition (fun lib -> allTargetProfiles = set lib.Targets )
-        
+            let sortedTargets = model.TargetsFileFolders |> List.sortBy (fun lib -> lib.Path)
+            //sortedTargets
+            //|> List.partition (fun lib -> allTargetProfiles = set lib.Targets )
+            ([] : LibFolder list), sortedTargets
```
Comment (matthid) [inline src/Paket.Core/Files/ProjectFile.fs]: This seems to be some code-path to "simplify" the generated project file. However it seems like it was never actually used.
After removing `runtime` this seems to work again (and breaks some tests).

Should we get rid of this simplification (as it was never used) or should I update the tests?

File: src/Paket.Core/InstallModel.fs [hunk 3]
```diff
@@ -115,171 +124,459 @@ type AnalyzerLib =
...
+        |> Option.orElseWith (fun _ ->
+            (trySscanf "runtimes/%A{rid}/native/%A{noSeperator}" p.PathWithinPackage : (Rid * string) option)
+            |> Option.map (fun (rid, _) -> { Path = Tfm.Empty; File = p; Runtime = rid }))
+
+    let getMsbuildFile (p:UnparsedPackageFile) =
+        (trySscanf "build/%A{tfm}/%A{noSeperator}" p.PathWithinPackage : (Tfm * string) option)
+        |> Option.map (fun (l,_) -> { Path = l; File = p; Runtime = RuntimeIdentifier.Any })
+        |> Option.orElseWith (fun _ ->
+            (trySscanf "build/%A{noSeperator}" p.PathWithinPackage : string option)
+            |> Option.map (fun (_) -> { Path = Tfm.Empty; File = p; Runtime = RuntimeIdentifier.Any }))
```
Comment (matthid) [inline src/Paket.Core/InstallModel.fs]: I think this part is now quite readable. I'm open for suggestions though. 
 - `trySscanf` is not completely type-safe (therefore the type annotations)
 - The `%A{scanner}` "extension" was invented by me to solve this problem

**Comment by matthid** [comment]
Related/Reverted https://github.com/fsprojects/Paket/pull/1626 and https://github.com/fsprojects/Paket/commit/a40e5d565b1a6c674fb555dec6ae45c3baacb16d

File: integrationtests/scenarios/i000346-aliases/before/MyClassLibrary/MyClassLibrary/MyClassLibrary.csprojtemplate [hunk 4]
```diff
@@ -1,7 +1,7 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
-  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
   <Import Project="..\..\packages\MultiTarget\build\MultiTarget.props" Condition="Exists('..\..\packages\MultiTarget\build\MultiTarget.props')" Label="Paket" />
+  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
```
Comment (matthid) [inline integrationtests/scenarios/i000346-aliases/before/MyClassLibrary/MyClassLibrary/MyClassLibrary.csprojtemplate]: Not quite sure about this change. 
What is the difference between a "global" and a "frameworkSpecific" msbuild file?
Paket seems to add global ones on top, but these tests seem to test the reverse?
Are those tests broken? What are the criteria for "global" msbuild?

Comment (0x53A) [inline integrationtests/scenarios/i000346-aliases/before/MyClassLibrary/MyClassLibrary/MyClassLibrary.csprojtemplate]: That was my try to fix https://github.com/fsprojects/Paket/issues/2227.
Nuget always adds props to the top. The distinction of "global" is specific to that, there is no such thing in nuget.

Comment (matthid) [inline integrationtests/scenarios/i000346-aliases/before/MyClassLibrary/MyClassLibrary/MyClassLibrary.csprojtemplate]: @0x53A thanks. so global is everything directly in the build folder, correct?
Is the current output acceptable? I have no Idea what I did that the ordering with the existing item changed... If not can you spot the relevant change (it's ok if not but than I know I have to take a deeper look)

**Comment by matthid** [comment]
I think this is ready to be reviewed or get feedback, though not completely finished.

I removed the runtime support as I think it is a cross cutting concern and therefore shouldn't be part of FrameworkIdentifier. 

Next step is to make runtime work again (but I think the big changes have been done).

@forki can you make an alpha out of this (leaving this pr open) so I can start testing on fake vnext?

File: integrationtests/scenarios/i000346-aliases/before/MyClassLibrary/MyClassLibrary/MyClassLibrary.csprojtemplate (hunk 4 above)
Comment (0x53A) [inline integrationtests/scenarios/i000346-aliases/before/MyClassLibrary/MyClassLibrary/MyClassLibrary.csprojtemplate]: @matthid I had tried to be smart. Global was everything either directly in the build folder, or for a list of frameworks that exactly matches the framework restriction. I think my use case still works (Fsharp.Compiler.Tools), I don't know if there are other problematic packages that Must be imported at the top.

Comment (matthid) [inline integrationtests/scenarios/i000346-aliases/before/MyClassLibrary/MyClassLibrary/MyClassLibrary.csprojtemplate]: @0x53A perfect, if you can live with this new behavior life is good

Comment (0x53A) [inline integrationtests/scenarios/i000346-aliases/before/MyClassLibrary/MyClassLibrary/MyClassLibrary.csprojtemplate]: For why it moved: previously it was (incorrectly) tagged as framework-specific, not global: 
image

With your simplified Partition(``List.partition (fun lib -> "" = lib.Path.Name )``), this is now correctly detected as global.

Comment (matthid) [inline integrationtests/scenarios/i000346-aliases/before/MyClassLibrary/MyClassLibrary/MyClassLibrary.csprojtemplate]: Thanks for confirming this!

**Comment by matthid** [comment]
Closing in favor of https://github.com/fsprojects/Paket/pull/2256

**Pull request closed without merging**